### DISCUSSION TITLE
Kubernetes Executor : Airflow command to run Task instance in new Pod container takes 7-8 seconds of time delay to start executing airflow command.

### DISCUSSION THREAD

**Discussion by vivek-zeta** [apache/airflow#23354] [Q&A]
Airflow Version : 2.2.2
Executor : Kubernetes Executor

As per my understanding after deep diving in to Airflow 2.2.2 code base.
When we run task in airflow it lunches new Kubernetes pod with airflow command as arguments to run task instance.
Sample command  args to the new Pod container : 
[
'`airflow`', '`tasks`', '`run`', '`dag_id`', '`task_id`', '`run_id`', '`--job-id'`, '`42748`', '`--raw`', '`--subdir`', '`DAGS_FOLDER/dags/dag_id.py`', '`--cfg-path`', '`/tmp/tmpzuvoil9g`', '`--error-file`', '`/tmp/tmp9mnrehws`'
]

- Through Airflow code testing and debugging with manually running various dummy task instance Pods,  I understood that 7-8 seconds of delay is coming before starting of the command execution. 
- It is confirmed that there are **no CPU throttling or any kind of resource laking** for running task instance in Kubernetes Pod container.
- May be time going in loading python libraries in parsing of the airflow command. Because with LocalExecutor we are not facing this delay as the task instance running as process in scheduler pod itself where libraries are already loaded to run any airflow commands. It's Just my view, please correct me if my understandings are not correct.
- I want to understand the root cause of why delay is coming in start of the airflow command?
- This delay is adding more than **1 Minute** in total execution time of the DAG Run.
- I hope someone with deep knowledge of airflow code base will help here to reduce this delay or remove it completely.

I'm open to address any question if anyone has in the problem statement
Please let me know in comments.

**Comment by vivek-zeta** [comment]
@potiuk  please check this 🙏🙏.

**Comment by jedcunningham** [comment]
What do you see when you run `time airflow version` in one of your pods?

Is 7s the delay between tasks you are seeing, via the UI? That seems roughly right with KubernetesExecutor. Keep in mind that you have the scheduler doing its thing, k8s getting the pod started, python startup time, then the task actually starting. This is the tradeoff of having separate workers with KubernetesExecutor, ultimately.

If you are sensitive to task delays, consider using CeleryExecutor or LocalExecutor instead. You could also use the combo executors, CeleryKubernetesExecutor, or new in 2.3.0 LocalKubernetesExecutor, so you can elect task-by-task which workers need isolation, and which tasks can run in a shared worker.

Lastly, if you haven't already increased it, it is highly recommended you increase [worker_pods_creation_batch_size](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#worker-pods-creation-batch-size), though normally that's the culprit when you are seeing higher task delays.

**Reply by vivek-zeta** [reply]
@jedcunningham 
Q : What do you see when you run time airflow version in one of your pods?
```
airflow@airflow-scheduler-1:/opt/airflow$ time airflow version
2.2.2

real	0m2.877s
user	0m2.688s
sys	0m0.173s
```

Q :  Is 7s the delay between tasks you are seeing, via the UI?
- In UI delay is `12-13` seconds.
- 7s delay is between task pod container started and actual command starts running.

-> worker_pods_creation_batch_size
- we already set it to 5. 
- but still facing 12-13 seconds delay in UI.

**Reply by potiuk** [reply]
> but still facing 12-13 seconds delay in UI.

Why do you think it's too much ? What is the average timeif creating a POD on your system? did you compare that?

**Reply by vivek-zeta** [reply]
@potiuk 4-5 seconds it takes to start the pod container. Remaining 7-8 seconds delay in starting of the airflow task run command.
So i want to reduce task command startup delay of 7-8 seconds.

**Reply by potiuk** [reply]
I think you simply have airflow code (and especially vendored in connexion importing is likely causing it ). At least somoene had very similar problem  in this discussion https://github.com/apache/airflow/discussions/21802. 

This is not the question of Airlfow and understanding it but - most likely understanding how to deploy any Python code in your cluster. If that's the case you likely wil have the same problem with any Python application because Python interpreter depends a lot on the fact that the python code it runs is on a relatively fast non-distributed filesysystem when it comes to importing a lot of librarires (Airlfow has ~600 dependencies). Especially if you know that by default when you start new container Python will try to import, parse and convert all the python files to bytecode and write the bytecode back so that next time it will work faster. There are two problems with that in container/k8s world: 

1) if you do not know that  (depending on your deployment) and have the source code of airflow on a slow (especially slow and synchronous  for write) filesystem then you can be bitten back

2) there is usually no "second time" so those .pyc and __pycache__ files written by Python interpreted will not be reused because containers are "run once only" by definition.

The `airflow version` taking 3 seconds is a clear indication of very, very slow import. When your Python code is on a very slow filesystem that has no optimisation and has synchronous writes, this is simply how long it takes airlfow to parse Python code and write back compiled code.

You can run `airflow version` in your container with py-spy as described here and send your .svg file here: https://github.com/apache/airflow/discussions/21802#discussioncomment-2246245 and see if you have similar results.

If you are interested in py-spy capabilities (pretty useful tool)- you can also try to analyse the performance yourself with py-spy following this great post from Ash https://www.astronomer.io/blog/profiling-the-airflow-scheduler/

I assume you have this problem (with slow filesystem where python code resides), then there are few ways you can approach it:

1) Absolute best - make sure filesystem your Airflow code is not on a slow filesystem. Review your deployment and make sure you only use local filesystem for Python code you use and no remote/slow filesystems are used.

2) set PYTHONDONOTWRITEBYTECODE variable to non-empty - this will avoid writing bytecode to filesystem (I suspect most of the problems are due to writing not reading). Writing is usually synchronous - so if you have slow/distributed filesystem where .pyc ad __pycache__ files are being written, the write operation done by python repeated multiple times and latency involved might compund to multiple seconds easily. This variable will turn Python sources in "read-only" and if writes are the problem, you might get significant boost this way without fixing the underlying problem with your filesystem,

3)  Upgrade to Airlfow 2.3.0 - there startup time will be better because of https://github.com/apache/airflow/pull/21438 - it will still take longer to start for more substantial commands though so this will only really help with "airflow version" startup time and few other commands, but some other commands might still need to import a lot of python code (but it might be distributed over longer period of time/ on demand and thus not as visible).

Generally speaking - this is your deployment problem, not Airflow.
